gst/mqtt/mqttsrc.c

```diff
@@ -409,12 +415,33 @@ gst_mqtt_src_class_finalize (GObject * object)
 {
   GstMqttSrc *self = GST_MQTT_SRC (object);
   GstBuffer *remained;
+  gboolean need_cleanup = FALSE;
 
+  /* Check if we need to cleanup MQTT client handle */
+  g_mutex_lock (&self->mqtt_src_mutex);
   if (self->mqtt_client_handle) {
-    MQTTAsync_destroy (&self->mqtt_client_handle);
-    self->mqtt_client_handle = NULL;
+    /** 
+     If stop wasn't called (element wasn't properly stopped),
+     we need to ensure safe cleanup here. */
+    if (GST_BASE_SRC_IS_STARTED (GST_BASE_SRC (self)) &&
+        self->is_connected && !self->is_disconnect_finished) {
+      need_cleanup = TRUE;
+    } else {
+      /* Safe to destroy immediately */
+      MQTTAsync_destroy (&self->mqtt_client_handle);
+      self->mqtt_client_handle = NULL;
+    }
+  }
+  g_mutex_unlock (&self->mqtt_src_mutex);
+
+  /* Perform disconnect outside of mutex lock to avoid deadlock */
+  if (need_cleanup) {
+    if (!gst_mqtt_src_disconnect (self, 2)) {
+      GST_WARNING_OBJECT (self, "Finalize MQTT disconnect failed, but continuing with cleanup");
+    }
   }
```

Comment:
@jaeyun-jung 
I've reflected your suggested code. I've also modified the finalize() code. Please check it.